Kei Finance - Defi Lama

kevupton opened the pull request:
**NOTE**

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. The protocol is usually listed within 24 hours of merging the PR
4. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
5. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
6. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
7. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Name (to be shown on DefiLlama):
Kei Finance

##### Twitter Link:
https://twitter.com/keifinance

##### List of audit links if any:


##### Website Link:
https://kei.fi/

##### Logo (High resolution, preferably in .svg and .png, for application on both white and black backgrounds. Will be shown with rounded borders):
https://assets.kei.fi/logo.png

##### Current TVL:
$589,688

##### Treasury Addresses (if the protocol has treasury)
0x3d027824a9eb4cc5e8f24d97fd8495ea9dc7026f

##### Chain:
Mainnet Ethereum

##### Coingecko ID (so your TVL can appear on Coingecko): (https://api.coingecko.com/api/v3/coins/list)
kei-finance

##### Coinmarketcap ID (so your TVL can appear on Coinmarketcap): (https://api.coinmarketcap.com/data-api/v3/map/all?listing_status=active,inactive,untracked&start=1&limit=10000)
24481

##### Short Description (to be shown on DefiLlama):
Kei Finance is an autonomous DeFi protocol powered by a suite of synchronised smart contracts. It offers reward mechanisms, treasury-backed floor prices, and staking solutions to its users.

##### Token address and ticker if any:
0xF75C7a59bCD9bd207C4Ab1BEB0b32EEd3B6392f3
https://etherscan.io/token/0xF75C7a59bCD9bd207C4Ab1BEB0b32EEd3B6392f3
(not sure what ticker means)

##### Category (full list at https://defillama.com/categories) *Please choose only one:
Liquid Staking
Staking
Services
Yield

##### Oracle used (Chainlink/Band/API3/TWAP or any other that you are using):


##### forkedFrom (Does your project originate from another project):


##### methodology (what is being counted as tvl, how is tvl being calculated):
TVL is based on how many tokens are currently locked in the staking contract. (They cannot be withdrawn until the staking committed has passed).

llamatester commented:
The adapter at projects/kei-finance exports TVL: 
        
 
 
    ethereum                  660.91 k
    
    total                    660.91 k

g1nt0ki commented:
@kevupton thanks for the PR, we can count the tokens in the treasury in the treasury metric but not as tvl, probably the staked Kei can be counted under 'staking' key, but I am unsure how to fetch this.

For more info: https://docs.llama.fi/list-your-project/what-to-include-as-tvl

kevupton commented:
Hey, @g1nt0ki sorry I am not 100% sure I am understanding, 
I had a read of the document, so I would assume it would fall under the staking category also,

g1nt0ki commented:
hi @kevupton would `totalValue()` call return the total amount of KEI staked?

kevupton commented:
Hey @g1nt0ki right now we are using the subgraph to track these fields, so would need to take the totalPrinciple + `totalRewards` where totalPrinciple is the accumulation of each stakes principle. Would we need to add a field on the contract for this to work for defillama? 

Also, another thing to note is the KEI token should be considered a reserve currency as it is backed by all the assets which are currently in the treasury (minus the tokens, and LP) which at this stage is just ETH.

g1nt0ki commented:
> Hey @g1nt0ki right now we are using the subgraph to track these fields, so would need to take the totalPrinciple + `totalRewards` where totalPrinciple is the accumulation of each stakes principle. Would we need to add a field on the contract for this to work for defillama?
> 
> Also, another thing to note is the KEI token should be considered a reserve currency as it is backed by all the assets which are currently in the treasury (minus the tokens, and LP) which at this stage is just ETH.

hi @kevupton , usage of subgraph is fine, but the KEI would only count towards staking, not tvl. Protocol controlled assets count towards treasury not tvl.

llamatester commented:
Error while running adapter at projects/kei-finance: 
        
 
  
    
    
    PromisePoolError: invalid address (argument="address", value="...", code=INVALID_ARGUMENT, version=address/5.7.0) (argument="account", value="...", code=INVALID_ARGUMENT, version=abi/5.7.0)
    Truncated error stack:
        at PromisePoolError.createFrom (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@supercharge/promise-pool/dist/promise-pool-error.js:28:16)
        at async Promise.all (index 0)

llamatester commented:
Error while running adapter at projects/kei-finance: 
        
 
  
    
    
    PromisePoolError: invalid address (argument="address", value="...", code=INVALID_ARGUMENT, version=address/5.7.0) (argument="account", value="...", code=INVALID_ARGUMENT, version=abi/5.7.0)
    Truncated error stack:
        at PromisePoolError.createFrom (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@supercharge/promise-pool/dist/promise-pool-error.js:28:16)
        at async Promise.all (index 0)

llamatester commented:
The adapter at projects/kei-finance exports TVL: 
        
 
 
    ethereum-staking          303.01 k
    ethereum                  50.60 k
    staking                   303.01 k
    
    total                    50.60 k

llamatester commented:
The adapter at projects/treasury/kei.js exports TVL: 
        
 
 
    ethereum-ownTokens        314.12 k
    ethereum                  226.08 k
    ownTokens                 314.12 k
    
    total                    226.08 k

kevupton commented:
@g1nt0ki I merged your branch into this one and updated the TVL and Staking in the index.js. 
Let me know if this looks better. I have removed the tokens from the TVL calculation.

g1nt0ki commented on projects/kei-finance/index.js:
```diff
@@ -0,0 +1,55 @@
+
+module.exports = {
+  ethereum: {
+    tvl,
```
change this to `tvl: () => 0,` then we are good, the value here is already counted under treasury, then the listing would be something like https://defillama.com/protocol/wonderland?treasury=true

llamatester commented:
The adapter at projects/kei-finance exports TVL: 
        
 
 
    ethereum                  0
    ethereum-staking          303.01 k
    staking                   303.01 k
    
    total                    0

llamatester commented:
The adapter at projects/treasury/kei.js exports TVL: 
        
 
 
    ethereum-ownTokens        314.12 k
    ethereum                  227.27 k
    ownTokens                 314.12 k
    
    total                    227.27 k

kevupton commented:
ok that has been updated now @g1nt0ki
